app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -8,72 +11,97 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
-          scope: 'users.omni_auth')
-      elsif is_navigational_format? && identity.errors.any?
-        # i18n-tasks-use t('users.omni_auth.failure_deauthorize')
-        set_flash_message(:alert, :failure_deauthorize, kind: OmniAuth::Utils.camelize(provider), scope: 'users.omni_auth',
-          reason: identity.errors.full_messages.join(', '))
+      if identity.nil?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize_not_linked')
+
+      elsif current_user.identities.count == 1 && !current_user.password_set?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize_last_identity')
+
+      elsif identity.destroy
+        remove_provider_from_session(omniauth_provider)
+        set_flash(:notice, 'users.omni_auth.success_deauthorize')
+
+      elsif identity.errors.any?
+        set_flash(:alert, 'users.omni_auth.failure_deauthorize', reason: identity.errors.full_messages.join(', '))
       end
       redirect_to edit_user_registration_path
     end
 
     private
 
+    def add_idp_to_existing_user # rubocop:disable Metrics/AbcSize
+      @current_user.identities << @user_identity
+      if @current_user.errors.blank?
+        store_location_for(:user, edit_user_registration_path)
+        set_flash(:notice, 'devise.omniauth_callbacks.success')
+        session['omniauth_provider'] = omniauth_provider
+        sign_in_and_redirect @current_user, event: :authentication
+      else
+        set_flash(:notice, 'devise.omniauth_callbacks.failure', reason: @current_user.errors.full_messages.join(', '))
+        redirect_to edit_user_registration_path
+      end
+    end
+
+    def sign_in_with_identity
+      user = @user_identity.user
+
+      # Update some profile information on every login if present
+      update_user_attributes(user)
+
+      sign_in_and_redirect user
+    end
+
+    def register_new_user # rubocop:disable Metrics/AbcSize
+      if omniauth_provider == 'nbp'
+        # go through NBP wallet connection process to create new account
+        session[:omniauth_provider_uid] = provider_uid
+        redirect_to nbp_wallet_connect_users_path
+      else
+        user = User.new_from_omniauth(omniauth_provider, provider_uid, omniauth_user_info)
+
+        if user.save
+          session['omniauth_provider'] = omniauth_provider
+          set_flash(:notice, 'devise.omniauth_callbacks.success')
+          sign_in_and_redirect user, event: :authentication
+        else
+          # Removing extra as it can overflow some session stores
+          session["devise.#{omniauth_provider}_data"] = request.env['omniauth.auth'].except('extra') # is this required?
+          set_flash(:alert, 'devise.omniauth_callbacks.failure', reason: user.errors.full_messages.join(', '))
+          redirect_to new_user_registration_url
+        end
+      end
+    end
+
+    def update_user_attributes(user)
+      user.assign_attributes(omniauth_user_info.slice(:email, :first_name, :last_name).to_h.compact)
```

Comment:
If a user identity is "dangling" to a deleted user, this will yield an exception.

<img width="1708" alt="Bildschirmfoto 2024-05-13 um 00 24 28" src="https://github.com/openHPI/codeharbor/assets/7300329/0163555d-45d1-43c0-88e6-d9b8f30692ca">

Reply:
Thats true but this _should_ never happen, right? Maybe we could add a database constraint though.

Reply:
Which constrain would that be? Since we are only soft-deleting the user, the referential integrity is still fine (at least regarding the ID). My suggestion: Fix the problem (you might have done so already), write a migration to remove "dangling" identities, and then forget about the issue. I would also agree that it _should_ never happen.

(Regarding the migration, [this one](https://github.com/openHPI/codeharbor/blob/master/db/migrate/20240312201146_migrate_task_descriptions_fix_kramdown_quotes.rb) might serve as an example. There are some things to consider, e.g., copying a minimized model to the migration, using `disable_ddl_transaction!`, using `in_batches`, ... A good read is [this Readme](https://github.com/fatkodima/online_migrations) for example on "backfilling data", even though we are not using the gem.)

Reply:
My mistake, I was thinking of enforcing referential integrity by using CASCADE on the users side but the user has no reference to the UserIdentities so this does not work.
Since the issue is already fixed, why do we need a migration?

Reply:
Couldn't be there some erroneous data in the production database that we should remove to restore the desired integrity?

Reply:
Actually I found out that the database is already enforcing the referential integrity (in schema.rb we add a foreign key from UserIdentity to User). I tested it using ```User.last.delete``` and the DB rejected the update. The reason why a UserIdentity can still have a ```nil``` user is that User has a ```default_scope``` which *also* gets applied when Rails fetches the User for a UserIdentity. This means that the soft-deletion causes the UserIdentities to be dangling. We can use the following to circumvent ```default_scope```:
```
class UserIdentity < ApplicationRecord
  belongs_to :user, -> { unscope(:where) }, inverse_of: :identities
```
It's not great but I think the only alternative would be to remove the ```default_scope``` which I think is not something for this PR.

Reply:
Yes, the `default_scope` is problematic here. Still, I would say we cannot simply remove the `default_scope` (for fetching the corresponding User for a given UserIdentity), since this would allow to undo the account deletion. Rather, we should really hard-delete the user identities when someone is deleting their account (similar as we deleted, e.g., messages). Then, we also don't need to `unscope` here. 

The migration I thought of could look like this (tested already, filename `20240526121908_destroy_dangling_user_identities.rb`):

```rb
# frozen_string_literal: true

class DestroyDanglingUserIdentities < ActiveRecord::Migration[7.1]
  def up
    UserIdentity.joins(:user).where(user: {deleted: true}).destroy_all
  end
end

class User < ApplicationRecord
end

class UserIdentity < ApplicationRecord
end
```

In the user model, we would need an new method being called within `handle_destroy`:

```rb
def handle_user_identities
  identities.destroy_all
end
```

I think, this should do the trick and resolve #1332 simultaneously.

Reply:
I copied your code :)